Sources/NIOSOCKS/Messages/ServerResponse.swift

```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+// MARK: - ServerResponse
+
+/// The SOCKS Server's response to the client's request
+/// indicating if the request succeeded or failed.
+public struct ServerResponse: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public var version: UInt8
+    
+    /// The status of the connection - used to check if the request
+    /// succeeded or failed.
+    public var reply: Reply
+    
+    /// The host address.
+    public var boundAddress: AddressType
+    
+    /// The host port.
+    public var boundPort: UInt16
+    
+    /// Creates a new `ServerResponse`.
+    /// - parameter reply: The status of the connection - used to check if the request
+    /// succeeded or failed.
+    /// - parameter boundAddress: The host address.
+    /// - parameter boundPort: The host port.
+    public init(reply: Reply, boundAddress: AddressType, boundPort: UInt16) {
+        self.version = 5
+        self.reply = reply
+        self.boundAddress = boundAddress
+        self.boundPort = boundPort
+    }
+    
+    init?(buffer: inout ByteBuffer) {
+        guard
+            let version = buffer.readInteger(as: UInt8.self),
+            let reply = Reply(buffer: &buffer),
+            buffer.readBytes(length: 1) == [0x00],
```

Comment:
Prefer `readInteger(as: UInt8.self)`, avoids a heap allocation.

Reply:
Also can I propose we store this into a `let` that has a name? I have no idea what this is.

Reply:
@Lukasa this is a "reserved" field used for some unknown reason. I can store some random byte "reserved" or something. Take a look at https://datatracker.ietf.org/doc/html/rfc1928 page 4 for an example.

Reply:
So let's just do `let reservedByte = buffer.readInteger(as: UInt8.self)`.

After all, we want to throw if this is set to an unexpected value.

Reply:
Sounds good